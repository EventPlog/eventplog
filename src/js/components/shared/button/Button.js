import React from 'react'
import styled, { css } from 'styled-components'
import colors from '../../../styles/theme/colors'
import { NavLink } from 'react-router-dom';
import { media } from 'js/styles/mixins'

const btnSizes = {
  medium: {fontSize: '1.2em'},
}
const commonStyles = css`
  border: 1px solid var(--activeLink);
  color: var(--activeLink);
  background: transparent;
  padding: 10px 30px;
  border-radius: 5px;
  text-transform: uppercase;
  font-size: 0.8rem;
  font-weight: 600;
  letter-spacing: 1px;
  
  ${
    media.phone`
      padding: 0.8rem;
    `
  }
  &:hover {
    background: var(--activeLink);
    color: ${colors.white};
  }
`

const inverted = css`
  background: var(--activeLink);
  color: ${ colors.white} !important;
  
  &:hover {
    background: ${ colors.white };
    color: var(--activeLink) !important;
  }
`

const StyledBtn = styled.button`
  ${ commonStyles }
`

const StyledLink = styled(NavLink)`
  ${ commonStyles }
  
  &:hover {
    color: ${colors.white} !important;
  }
`;

const InvertedBtn = (Component, props) => (
  props.inverted
  ? Component.extend`
      ${ inverted }
    `
  : Component
)

const Button = function(props) {
  const Component = InvertedBtn(StyledBtn, props)
  return <Component {...props} />
}

Button.Link = function(props) {
  const Component = InvertedBtn(StyledLink, props)
  return <Component {...props} />
}


export default Button
