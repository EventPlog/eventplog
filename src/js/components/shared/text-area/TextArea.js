import React from 'react'
import styled, { css } from 'styled-components'
import {TextArea} from 'semantic-ui-react'

const styles = css`
  flex: 1;
  border: 1px solid #ddd;
  border-radius: 10px;
  padding: 10px;
  
  &::placeholder {
    color: #ccc;
  }
`

const TextAreaComponent = React.forwardRef(({
  className,
  placeholder,
  children,
  ...otherProps
}, ref) => (
  <TextArea ref={ref} className={className} placeholder={placeholder} {...otherProps}>
    {children}
  </TextArea>
))
export default styled(TextAreaComponent)`${styles}`


