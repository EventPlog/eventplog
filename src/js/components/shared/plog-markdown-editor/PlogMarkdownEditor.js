import React from 'react'
import { Icon } from 'semantic-ui-react'

// internal components
import TextArea from 'js/components/shared/text-area'
import Button from 'js/components/shared/button'
import styled from 'styled-components'
import { splice } from 'js/utils'


const TextEditorStyles = styled.div`
  textarea {
    min-height: 200px;
  }
  
  .toolbox {
    display: flex;
    flex-wrap: wrap;
    
    > * {
      padding: 0.5rem 1rem;
      border-bottom: none;
      cursor: pointer;
      
      &:hover {
        background: ${props => props.theme.gray};
      }
    }
    
    .bold {
      font-weight: 800;
    }
    
    .italic {
      font-style: italic;
    }
    
    .underline {
      text-decoration: underline;
    }
  }
  
  .save-btn {
    padding: 0.5rem 1rem;
    margin: 0.5rem 1rem 0.5rem 0;
    
  }
`

export const format = (preCh, postCh, state, convertBr) => {
  const { selectionStart, selectionEnd, selectedText } = state;

  // if (selectionStart == selectionEnd) return

  let textValue = state.textValue;
  textValue = splice(selectionEnd, 0, textValue, postCh)
  if (convertBr) {
    textValue = splice(selectionStart, selectedText.length, textValue, selectedText.replace(/(\r?\n)(\s+)?/g, convertBr))
  }
  textValue = splice(selectionStart, 0, textValue, preCh)
  return textValue
}

class PlogMarkdownEditor extends React.Component {
  constructor(props) {
    super();
    this.state = {}
  }

  componentWillMount() {
    this.setState({ textValue: this.props.value })
  }

  format = (preCh, postCh, convertBr) => {
    const textValue = format(preCh, postCh, this.state, convertBr)
    this.setState({ textValue })
    this.saveChanges(textValue)
  }

  onClick = (el, att) => {
    const {selectionStart, selectionEnd } = el.target
    const selectedText = el.target.textContent
                            .substr(selectionStart, selectionEnd - selectionStart)
    this.setState({ selectionStart, selectionEnd, selectedText })
  }

  saveChanges(value) {
    this.props.onChange({target: {value}})
  }

  onChange = (el, attr) => {
    this.setState({ textValue: attr.value })
    this.props.onChange(el, attr)
  }

  render() {
    const lineBreaks = `

`
    const lineBreak = `
`
    return (
      <TextEditorStyles>
        <div className="toolbox">
          <div className="cell"
               onClick={() => this.format('#### ', '')}>
            <Icon name="heading" />
          </div>
          <div className="cell bold"
               onClick={() => this.format('**', '**')}>
            <Icon name="bold" />
          </div>
          <div className="cell italic"
               onClick={() => this.format('*', '*')}>
            <Icon name="italic" />
          </div>
          <div className="cell"
               onClick={() => this.format('__', '__')}>
            <Icon name="underline" />
          </div>
          <div className="cell"
               onClick={() => this.format('~~', '~~')}>
            <Icon name="strikethrough" />
          </div>
          <div className="cell paragraph"
            onClick={() => this.format(lineBreaks, '')}>
            <Icon name="paragraph" />
          </div>
          <div className="cell"
               onClick={() => this.format('> ', lineBreak)}>
            <Icon name="quote left" />
          </div>
          <div className="cell list"
               onClick={() => this.format(`${lineBreak}1. `, lineBreak, `${lineBreak}1. `)}>
            <Icon name="list ol" />
          </div>
          <div className="cell"
               onClick={() => this.format(`${lineBreak}* `, lineBreak, `${lineBreak}* `)}>
            <Icon name="list ul" />
          </div>
          <div className="cell code"
               onClick={() => this.format('`', '`')}>
            <Icon name="code" />
          </div>
          <div className="cell"
            onClick={() => this.format('```' + lineBreak, lineBreak + '```' + lineBreak)}>
            <Icon name="file code outline" />
          </div>
          <div className="cell"
               onClick={() => this.format('![file_name]', '(replace_with_image_url)')}>
            <Icon name="image" />
          </div>
        </div>
        <TextArea {...this.props}
                  onBlur={() => {} }
                  onKeyUp={this.onClick}
                  onClick={this.onClick}
                  value={this.state.textValue}
                  onChange={this.onChange} />
        {this.props.showSubmit &&
          <span>
            <Button className="save-btn"
                    inverted
                    onClick={this.props.onSubmit}>
              <Icon name="save" /> Save
            </Button>
            {this.props.onCancel &&
              <Button className="save-btn"
                      onClick={this.props.onCancel}>
                <Icon name="save"/> Cancel
              </Button>
            }
          </span>
        }
      </TextEditorStyles>
    )
  }
}

export default PlogMarkdownEditor

