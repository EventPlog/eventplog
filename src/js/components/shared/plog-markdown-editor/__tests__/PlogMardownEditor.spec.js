import React from 'react'
import PlogMarkdownEditor, {
  format
} from '../PlogMarkdownEditor'
import { shallow } from 'enzyme'
import TextArea from 'js/components/shared/text-area'
import Button from 'js/components/shared/button'

describe('PlogMarkdownEditor', () => {
  const props = {
    value: 'Everyday sun',
    onChange: () => {},
    showSubmit: true,
    onSubmit: () => {},
  }
  const wrapper = shallow( <PlogMarkdownEditor {...props} /> );

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()

    expect(wrapper.find(TextArea).length).toEqual(1)
    expect(wrapper.find(Button).length).toEqual(1)
  })
})

describe('format', () => {
  const text = 'This is a text'
  const state = {
    selectionStart: 'This '.length,
    selectionEnd:  'This is'.length,
    selectedText: 'is',
    textValue: text,
  }

  it('appends the right pre and post characters to a str', () => {
    expect(format('**', '**', state)).toEqual('This **is** a text')
  })

  describe('when a linebreak char is supplied', () => {
    const lineBreak = `
`
    const lineBreaks = `

`
    const str = `
There is no time for
melodrama
parental care
Or anything like that
`
    const selectionStart = `
There is no time for
`.length

    const selectionEnd = `
There is no time for
melodrama
parental care`.length

    const selectedText = `melodrama
parental care`

    const expected = `
There is no time for

* melodrama
* parental care

Or anything like that
`

    const newState = {
      selectionStart, selectionEnd, selectedText, textValue: str
    }
    const result = format(`${lineBreak}* `, lineBreak, newState, `${lineBreak}* `)
    it('replaces line breaks with the supplied char', () => {
      expect(result).toEqual(expected)
    })
  })
})