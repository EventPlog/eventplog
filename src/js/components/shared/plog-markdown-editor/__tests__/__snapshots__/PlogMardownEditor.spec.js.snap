// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlogMarkdownEditor renders correctly 1`] = `
<styled.div>
  <div
    className="toolbox"
  >
    <div
      className="cell"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="heading"
      />
    </div>
    <div
      className="cell bold"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="bold"
      />
    </div>
    <div
      className="cell italic"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="italic"
      />
    </div>
    <div
      className="cell"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="underline"
      />
    </div>
    <div
      className="cell"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="strikethrough"
      />
    </div>
    <div
      className="cell paragraph"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="paragraph"
      />
    </div>
    <div
      className="cell"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="quote left"
      />
    </div>
    <div
      className="cell list"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="list ol"
      />
    </div>
    <div
      className="cell"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="list ul"
      />
    </div>
    <div
      className="cell code"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="code"
      />
    </div>
    <div
      className="cell"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="file code outline"
      />
    </div>
  </div>
  <Styled(Component)
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onKeyUp={[Function]}
    onSubmit={[Function]}
    value="Everyday sun"
  />
  <Button
    className="save-btn"
    onClick={[Function]}
  >
    <Icon
      as="i"
      name="save"
    />
     Save
  </Button>
</styled.div>
`;
